packages/cube/lib/cube/buildCubeShape/constraintBuilder/DimensionConstraintsBuilder.js

```diff
@@ -65,6 +69,17 @@ export class DimensionConstraintsBuilder {
       builders[0].build(ptr)
     }
     if (builders.length > 1) {
+      // if all builders have sh:in, then merge them
+      if (builders.every(hasValuesBuilder)) {
+        const merged = new ValuesConstraintBuilder(this.rdf)
+        for (const compositeBuilder of builders) {
+          const valuesBuilder = getValuesBuilder(compositeBuilder)
+          valuesBuilder.enabled = false
+          valuesBuilder.values.forEach(value => merged.add(value))
+        }
+        merged.build(ptr)
+      }
```

Comment:
Maybe it would be a little more readable if you'd simply loop the builders and `break` when any one of them does not have values builder

Trades `hasValuesBuilder` for `disableValuesBuilder` and reduces nesting. Up to you

```suggestion
      let merged = new ValuesConstraintBuilder(this.rdf)
      for (const compositeBuilder of builders) {
        const valuesBuilder = getValuesBuilder(compositeBuilder)
        if (!valuesBuilder) {
          merged = null
          break
        }

        valuesBuilder.values.forEach(value => merged.add(value))
      }
      if (merged) {
        builders.forEach(disableValuesBuilder)
        merged.build(ptr)
      }
```

Reply:
what about instead:

```js
      const valuesBuilders = builders.map(getValuesBuilder)
      if (valuesBuilders.every(builder => builder && builder.enabled)) {
        const merged = new ValuesConstraintBuilder(this.rdf)
        const allValues = valuesBuilders.flatMap(builder => Array.from(builder.values))
        allValues.forEach(value => merged.add(value))
        valuesBuilders.forEach(builder => { builder.enabled = false })
        merged.build(ptr)
      }
```

Reply:
Compact! :+1:

Reply:
Does this work?

```js
if (valuesBuilders.every(builder => builder?.enabled)) {
```